### PR TITLE
As a D4R dev, I want D4R UI nodes to deserialize w/o adding duplicate ports

### PR DESCRIPTION
**Pull request by alfarok**
### Purpose

[Task: QNTM-1708](https://jira.autodesk.com/browse/QNTM-1708)

[CoincidingD4R Changes ](https://github.com/DynamoDS/DynamoRevit/pull/1783)
(Dynamo Beta NuGet packages must be updated for successful Jenkins build of D4R)

This PR addresses the duplicate port issue during deserialization.  This is related to the recent changes in the node model base class affecting all the Revit UI nodes that inherent from this base.  Json constructors must be added throughout the inheritance chain to prevent this from occurring.

### Testing
Added new regression testing to verify no nodes are lost in translation after saving/reopening and that port counts are consistent.  The tests:
- iterate through specified assemblies in the library, adding all those nodes to the canvas
- store specific data (active node count, port count, etc)
- save temp file, close, reopen
- verify data is consistent with original state after serialization/deserialization
- close/delete temp file

No corresponding dyn is required reducing testing maintenance.

### Affected Nodes (49):
**[Elements.cs]**
- All Elements of Family Type
- All Elements of Type
- All Elements of Category
- All Elements at Level
- All Elements in Active View

**[RevitDropDown.cs]**
- Family Types
- Get Family Parameter
- Floor Types
- Wall Types
- Performance Adviser Rules
- Categories
- Levels
- Structural Framing Types
- Spacing Rule Layout
- Element Types
- Views

**[RevitTypes.cs]**
- Select Phase
- Select Revision
- Select Filled Region Type
- Select Rule Type
- Select Revision Numbering
- Select Revision Numbering Type
- Select Parameter Type
- Select BuiltIn Parameter Group
- Select Revision Visibility
- Select Direct Shape Room Bounding Option
- Detail Level
- Select Horizontal Text Alignment
- Select Vertical Text Alignment
- Wall Location
- Schedule Type
- Export Column Headers
- Export Text Qualifier
- Fill Patterns
- FIll Pattern Targets
- Line Patterns
- Schedule Filter Type

**[Selection.cs]** (these were addressed in [1768](https://github.com/DynamoDS/DynamoRevit/pull/1768))
- Select Model Element
- Select Face
-  Select Edge
- Select Point on Face
- Select UV on Face
- Select Divided Surface Families
- Select Model Elements
- Select Faces
- Select Edges

**[SiteLocation.cs]**
- Site Location

**[SunPath.cs]**
- SunSettings Current

### Outstanding Issues
- Resolve `SiteLocation` Deserialization Issues (verified this is old behavior in 1.3 and filed [QNTM-1903](https://jira.autodesk.com/browse/QNTM-1903))
- Node State is inconsistent pre/post save (verified this is also old behavior in 1.3 and the behavior is consistent with the existing 2.0 core custom UI nodes(filed [QNTM-1902](https://jira.autodesk.com/browse/QNTM-1902) with more info/examples)
### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] Snapshot of UI changes, if any.

### Reviewers

@mjkkirschner @QilongTang

### CODE REVIEW

File: src/Libraries/CoreNodeModels/DropDown.cs [hunk 1]
```diff
@@ -97,7 +97,7 @@ protected DSDropDownBase(string outputName)
         }
 
         [JsonConstructor]
-        protected DSDropDownBase(IEnumerable<PortModel> inPorts, IEnumerable<PortModel> outPorts) : base(inPorts, outPorts)
+        protected DSDropDownBase(string outputName, IEnumerable<PortModel> inPorts, IEnumerable<PortModel> outPorts) : base(inPorts, outPorts)
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodeModels/DropDown.cs]: whats this change about?

File: src/Libraries/CoreNodeModels/Enum.cs [hunk 2]
```diff
@@ -12,6 +12,11 @@ namespace CoreNodeModels
 {
     public abstract class EnumAsInt<T> : EnumBase<T>
     {
+        protected EnumAsInt() { }
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodeModels/Enum.cs]: please add a comment where you needed to add these.

File: src/Libraries/CoreNodeModels/Enum.cs [hunk 3]
```diff
@@ -61,7 +69,7 @@ protected AllChildrenOfType() : base("Types")
         }
 
         [JsonConstructor]
-        protected AllChildrenOfType(IEnumerable<PortModel> inPorts, IEnumerable<PortModel> outPorts) : base(inPorts, outPorts)
+        protected AllChildrenOfType(IEnumerable<PortModel> inPorts, IEnumerable<PortModel> outPorts) : base("Types", inPorts, outPorts)
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodeModels/Enum.cs]: I think where possible we should start pulling strings out of the constructors and creating constants we can reference.

File: src/Libraries/CoreNodeModels/DropDown.cs (hunk 1 above)
Comment (alfarok) [reply src/Libraries/CoreNodeModels/DropDown.cs]: if you look at the non-json constructor above it takes a single string parameter for the `outputName`.  The json constructor was previously created in another PR but this parameter was missing and is required to correctly pass the `outputName` from the child to the base constructor

File: src/Libraries/CoreNodeModels/Enum.cs (hunk 2 above)
Comment (alfarok) [reply src/Libraries/CoreNodeModels/Enum.cs]: see [this comment](https://github.com/DynamoDS/DynamoRevit/pull/1783#discussion_r138064533) in the D4R portion of the PR

**Pull request merged**